server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -401,7 +402,11 @@ private FrameBoundDefinition convertToAnalyzedFrameBound(ExpressionAnalysisConte
         int size = symbolsToCast.size();
         List<Symbol> castList = new ArrayList<>(size);
         for (int i = 0; i < size; i++) {
-            castList.add(symbolsToCast.get(i).cast(targetTypes.get(i), false));
+            var symbolToCast = symbolsToCast.get(i);
+            if (ArrayType.unnest(symbolToCast.valueType()).id() != ObjectType.ID) {
+                symbolToCast = symbolToCast.cast(targetTypes.get(i), false);
+            }
```

Comment:
Why are arrays of objects skipped here? Shouldn't the array-type/object-type be adapted instead?

Reply:
The description of this point from the chat communication 

```
The second issue is related to subscript on object or arrays of objects. There we try to infer the return type of the function at the normalization step based on arguments, see https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/expression/scalar/SubscriptObjectFunction.java#L98, but with the new function signature, the object arguments were cast to objects coz of different inner types (empty vs actual inner types) and the return type inference logic was not working any longer.
```

so basically I did that to skip cast for `object` and `array(object)` i am not sure how to handle in another way. hmm maybe actually checking whether the source and target are either `array(object)` and `object` and then don't cast source type if the source and target types have different inner types?